Game/MyProject/Game/ModelLoader/ModelLoader.cpp

```diff
@@ -0,0 +1,129 @@
+//
+//  ModelLoader.cpp
+//  Game
+//
+//  Created by Tamas Both - (p) on 09/04/2019.
+//
+
+#include "ModelLoader.hpp"
+
+std::string ModelLoader::directory;
+IvTexture*  ModelLoader::texture;
+
+// --------------------------------
+// PUBLIC FUNCTION(S) AND METHOD(S)
+// --------------------------------
+
+// loads a model into the scene graph
+// the given scene_node parameter will be the parent of the model
+void ModelLoader::loadModel(const std::string& path, const std::string& texture_path, const std::string& shader_path, SceneNode* scene_node) {
+    Assimp::Importer            import;
+    const aiScene*              new_scene;
+    std::shared_ptr<SceneNode>  new_root;
+    
+    new_scene = import.ReadFile(path, aiProcess_Triangulate | aiProcess_FlipUVs);
+    if(!new_scene || new_scene->mFlags & AI_SCENE_FLAGS_INCOMPLETE || !new_scene->mRootNode) {
+        std::cout << "ERROR::ASSIMP::" << import.GetErrorString() << std::endl;
+        return;
+    }
+    
+    directory = path.substr(0, path.find_last_of('/'));
+    
+    new_root = std::make_shared<SceneNode>(new_scene->mRootNode->mName.data);
+    scene_node->addChild(new_root);
+    
+    loadTexture(texture_path.c_str());
+    processNode(new_scene->mRootNode, new_scene, new_root.get(), shader_path);
+}
+
+// --------------------------------
+// PRIVATE FUNCTION(S) AND METHOD(S)
+// --------------------------------
+
+// transforms the assimp scene node to my scene node
+void ModelLoader::processNode(aiNode* node, const aiScene* scene, SceneNode* my_scene_node, const std::string& shader_path) {
+    for (unsigned int i = 0; i < node->mNumMeshes; i++) {
+        aiMesh *mesh = scene->mMeshes[node->mMeshes[i]];
+        my_scene_node->addChild(makeSceneNode(mesh, scene, node, shader_path));
+    }
+    
+    for (unsigned int i = node->mNumMeshes; i < node->mNumChildren; i++)
+        my_scene_node->addChild(makeSceneNode(nullptr, scene, node, shader_path));
+    
+    for (unsigned int i = 0; i < node->mNumChildren; i++)
+        processNode(node->mChildren[i], scene, my_scene_node->getChild(i), shader_path);
+}
+
+// returns a new scene node
+std::shared_ptr<SceneNode> ModelLoader::makeSceneNode(const aiMesh *mesh, const aiScene *scene, const aiNode* node, const std::string& shader_path) {
+    std::shared_ptr<Mesh>           my_mesh;
+    std::shared_ptr<MeshInstance>   mesh_instace;
+    std::shared_ptr<SceneNode>      scene_node;
+    
+    std::vector<IvTCPVertex>        vertices;
+    std::vector<unsigned int>       indices;
+    
+    aiVector3t<float>               scale;
+    aiVector3t<float>               rotation;
+    aiVector3t<float>               position;
+    
+    scene_node = std::make_shared<SceneNode>(node->mName.data);
+    
+    if (mesh) {
+        for (unsigned int i = 0; i < mesh->mNumVertices; i++) {
+            IvTCPVertex vertex;
```

Comment:
Ok for now but you might have to use generic vertex format. It is not uncommon to have multiple textures uvs, tangents, normals, etc. This can be done using the `VertexDescription` class